File: idioms/python/ConstantPropertyByClosure.py
#! /usr/bin/env python

import copy
import pprint
import types
import sys

def ConstantProperty(value):
    """Create a closure protected constant copy of value."""
    cache = copy.deepcopy(value)
    return property(
        lambda self: copy.deepcopy(cache)
    )

class A(object):
    def __init__(self, value):
        """Class constructor."""
        A.constant = ConstantProperty(value)

    def set_constant(self, value):
        """A non property setter for constant."""
        self.constant = value

mutable = {
    'A': [1, 2],
    'B': 'cannot be change'
}

# Initialize object with a mutable object
a = A(mutable)
# Print object
pprint.pprint(a.constant)

# Try to change constant property
# by updating object referred by "value"
mutable['B'] = 'can be changed'
pprint.pprint(a.constant)

# Try to change constant property
# by updating object return by a.constant
a.constant['B'] = 'can be changed'
pprint.pprint(a.constant)

# Try to modify constant property
try:
    a.constant = { 'A': 'new' }
    pprint.pprint(a.constant)
except Exception as error:
    print(error)

# Try to modify constant property
# from a class member function
try:
    a.set_constant({ 'A': 'new' })
    pprint.pprint(a.constant)
except Exception as error:
    print(error)

File: idioms/javascript/ConstantPropertyByClosure.js
#! /usr/bin/env node
"use strict";

// Create a closure protected constant copy of value.
function ConstantProperty(object, name, value){
  const cache = JSON.parse(
    JSON.stringify(value)
  );
  Object.defineProperty(object, name, {
    get: () => JSON.parse(
      JSON.stringify(cache)
    )
  });
}

class A {
  // Class constructor
  constructor(value) {
    ConstantProperty(this, 'constant', value);
  }
  // A non property setter for constant.
  setConstant(value) {
    this.constant = value;
  }
}

const mutable = {
    'A': [1, 2],
    'B': 'cannot be change'
};

// Initialize object with a mutable object
const a = new A(mutable);
// Print object
console.log(a.constant);

// Try to change constant property
// by updating object referred by mutable
mutable['B'] = 'can be changed';
console.log(a.constant);

// Try to change constant property
// by updating object return by a.constant
a.constant['B'] = 'can be changed';
console.log(a.constant);

// Try to modify constant property
try {
    a.constant = { 'A': 'new' };
    console.log(a.constant);
} catch(error) {
    console.log(error);
}

// Try to modify constant property
try {
    a.setConstant({ 'A': 'new' });
    console.log(a.constant);
} catch(error) {
    console.log(error)
}
